ArduPlane/mode_autoland.cpp

```diff
@@ -0,0 +1,118 @@
+#include "mode.h"
+#include "Plane.h"
+#include <GCS_MAVLink/GCS.h>
+
+#if MODE_AUTOLAND_ENABLED
+
+/*
+  mode AutoLand parameters
+ */
+const AP_Param::GroupInfo ModeAutoLand::var_info[] = {
+    // @Param: WP_ALT
+    // @DisplayName: Final approach WP altitude
+    // @Description: This is the target altitude above HOME for final approach waypoint
+    // @Range: 0 200
+    // @Increment: 1
+    // @Units: m
+    // @User: Standard
+    AP_GROUPINFO("WP_ALT", 1, ModeAutoLand, final_wp_alt, 55),
+
+    // @Param: WP_DIST
+    // @DisplayName: Final approach WP distance
+    // @Description: This is the distance from Home that the final approach waypoint is set. The waypoint point will be in the opposite direction of takeoff (the direction the plane is facing when the plane sets its takeoff heading)
+    // @Range: 0 700
+    // @Increment: 1
+    // @Units: m
+    // @User: Standard
+    AP_GROUPINFO("WP_DIST", 2, ModeAutoLand, final_wp_dist, 400),
+
+    // @Param: DIR_OFF
+    // @DisplayName: Landing direction offset from takeoff
+    // @Description: The captured takeoff direction (at arming,if TKOFF_OPTION bit1 is set, or after ground course is established in autotakeoffs)is offset by this amount to create a different landing direction and approach.
+    // @Range: -360 360
+    // @Increment: 1
+    // @Units: deg
+    // @User: Standard
+    AP_GROUPINFO("DIR_OFF", 3, ModeAutoLand, landing_dir_off, 0),
+
+    AP_GROUPEND
+};
+
+ModeAutoLand::ModeAutoLand() :
+    Mode()
+{
+    AP_Param::setup_object_defaults(this, var_info);
+}
+
+bool ModeAutoLand::_enter()
+{   
+    //must be flying to enter
+    if (!plane.is_flying()) { 
+        gcs().send_text(MAV_SEVERITY_WARNING, "Must already be flying!"); 
+        return false;
+    }
+    
+    //takeoff direction must be set and must not be a quadplane, otherwise since flying switch to RTL so this can be used as FS action
+#if HAL_QUADPLANE_ENABLED
+    if (quadplane.available() && !quadplane.option_is_set(QuadPlane::OPTION::ALLOW_FW_LAND)) {
+        gcs().send_text(MAV_SEVERITY_WARNING, "Option not set to allow fixed wing autoland"); 
+        return false; 
+    }
+#endif
+    if (!plane.takeoff_state.takeoff_direction_initialized) {
+        gcs().send_text(MAV_SEVERITY_WARNING, "Takeoff initial direction not set,must use autotakeoff");
```

Comment:
This message is a bit long. I'm sure it's going to wrap on Mission Planner. Make two messages?

Reply:
it will wrap in the MP messages tab, like many other messages...but making it two will make the first one not show in HUB and OSD where last GCS send text always wins

Reply:
Yes, lets not try and fix Mission Planner issues in AP. It should re-assemble messages if it needs to.